### PR TITLE
Use new TTGlyphPointPen from fonttools

### LINKED ISSUE #469: TrueType outline compiler may change point indices
The TrueType outline compiler changes the point indices in some cases.

https://github.com/googlefonts/ufo2ft/blob/master/Lib/ufo2ft/outlineCompiler.py#L1301-L1316

This is because it uses a segment-based pen, which rotates the points so each contour starts with an oncurve point.

This is a problem for instructed glyphs, as it makes the instructions invalid. Using a point-based pen could solve the problem, but it doesn't exist in fontTools.

### PR DESCRIPTION
**Pull request by jenskutilek**
...  to avoid modification of TT outlines. Fixes #469.

### CODE REVIEW

**Comment by jenskutilek** [comment]
Depends on https://github.com/fonttools/fonttools/pull/2205.

**Comment by anthrotype** [comment]
we need to bump minimum required fonttools in setup.py

**Comment by anthrotype** [comment]
AttributeError: StubGlyph has no attribuet "drawPoints"

https://github.com/googlefonts/ufo2ft/pull/478/checks?check_run_id=2608944622#step:5:1509

Some tests would be nice

**Comment by jenskutilek** [comment]
What do you suggest should be tested? (which isn't covered by fonttools/TTGlyphPointPen or ufo2ft [TrueType compiler branch](https://github.com/googlefonts/ufo2ft/pull/335))

**Comment by anthrotype** [comment]
You didn't need to change tests expectations after this patch (good, it means it continues to work) because none of the test ufo's glyphs contain off-curve points as starting points.
How about, in outlineCompiler_test::OutlineTTFCompilerTest, you add a test that verifies that point list is not rotated even if a contour starts with off-curve point; e.g. you add a new glyph to the testufo (within the test code itself, not necessarily in the actual testufo on disk as that may affect other tests that use the same fixture) with a contour starting with an off-curve point and assert that the first point on the TTGlyph is still the same off-curve point.

**Comment by anthrotype** [comment]
do you mind also running black to fix the CI lint step. It's unrelated to your change (apparently black got pickier about docstrings), but it doesn't hurt to do it here

**Comment by anthrotype** [comment]
hm that's a lot of files. also not clear very mich what "test_instructions" is really testing.. I was thinking more of you take the default testufo, add a glyph with starting off-curve (in the test python code itself, before you do the asserts) then check that the compiled glyph is as expected. A more tailored test, in other words. This one is basically like a whole intergration test

File: tests/integration_test.py [hunk 1]
```diff
@@ -237,6 +242,19 @@ def test_drop_glyph_names(self, testufo, output_format, options, expected_ttx):
...
+    def test_Instructions(self, instructions_ufo):
+        ttf = compileTTF(
+            instructions_ufo,
+            convertCubics=False,
+            reverseDirection=False,
+            removeOverlaps=False
+        )
+        assert "cvt " not in ttf
+        assert "gasp" in ttf
+        assert "fpgm" not in ttf
```
Comment (anthrotype) [inline tests/integration_test.py:254]: won't these asserts have to change after we merge the PR with hinting support?

**Comment by anthrotype** [comment]
we'll definitely need an integration test for https://github.com/googlefonts/ufo2ft/pull/335, but for this I would like a unit test in outlineCompiler_test.py that checks that the resulting off-curve-starting glyph from OutlineCompiler is as expected

**Comment by anthrotype** [comment]
so annoying github wants me to approve the CI running each time you push a commit, otherwise they won't run automatically. I think it's to prevent bitcoin mining from first-time contributor. Oh well, I just made you a collaborator with write access, let's see if that works

**Comment by jenskutilek** [comment]
I have adapted a quadratic test UFO that will be added by #335 anyway. Here the compiled font has no hinting though, but in glyph `o` there is the offcurve start point that must be preserved.

It is very tricky, because even calling the preProcessor (in particular the `CubicToQuadraticFilter`) on an already quadratic UFO will change the start point to the next oncurve point. (This cost me a couple of debugging hours this morning...)

So testing the OutlineCompiler may not be enough.

File: tests/integration_test.py (hunk 1 above)
Comment (jenskutilek) [reply tests/integration_test.py:254]: Yes, this is a test I adapted from #335. The asserts must be inverted there.

**Comment by anthrotype** [comment]
> calling the preProcessor (in particular the CubicToQuadraticFilter) on an already quadratic UFO will change the start point

yes, you want to disable that filter if you have a UFO with quadratic curves, there's an option for it

> testing the OutlineCompiler may not be enough.

well, it tests that OutlineCompiler doesn't mess with the points anymore

**Comment by anthrotype** [comment]
> This cost me a couple of debugging hours this morning

sorry about that. yes, quadratic ufos should turn avoid passing through cu2qu (which probably uses segment pen protocol so it rotates)  and also disable reversing contour direction; if hinted, you should also disable removeOverlaps. But you figured that out already

**Comment by jenskutilek** [comment]
> so annoying github wants me to approve the CI running each time you push a commit, otherwise they won't run automatically. I think it's to prevent bitcoin mining from first-time contributor. Oh well, I just made you a collaborator with write access, let's see if that works

Thanks, it works now :)

**Comment by jenskutilek** [comment]
I think this is ready to merge. Any objections @anthrotype?

**Comment by anthrotype** [comment]
i still think it's unnecessary to add such a huge set of test files just for testing this relatively small change from TTGlyphPen to TTGlyphPointPen, so I cherry picked your relevant commits and added a smaller test in #503 - please take a look and say "@googlebot I consent." if you're happy with me committing something on your behalf
